### ISSUE TITLE
Support for ARCore

### ISSUE THREAD

**Issue by rugk** [microg/GmsCore#1672]
**Is your feature request related to a problem? Please describe.**
[ARCore](https://arvr.google.com/arcore/) is an quite oldish (2017-released) already [SDK](https://github.com/google-ar/arcore-android-sdk/) for doing AR/VR related things like measuring via your camera.

Now [StreetComplete is working on integrating it](https://github.com/streetcomplete/StreetComplete/pull/3709#issuecomment-1039712252).

The problem is it depends on Google Play Services/a proprietary app as far as I see.

**Describe the solution you'd like**
Of course, it would be awesome if one could make it work:
* if one chooses to install the proprietary ARCore app
* or even without the ARCore and only microG installed

**Describe alternatives you've considered**
I know especially the second solution is very likely a huge and quite impossible task for this project (or are FLOSS AI models and tensorflow stuff already enough to implement such a thing?), so that's why I also looked [into your wiki status page](https://github.com/microg/GmsCore/wiki/Implementation-Status) but could not find that feature listed there.

As such, the alternative would be to document the (intended) support status at least.

**Additional context**
* Even Google only supports selected devices: https://developers.google.com/ar/devices
* Android docs: https://developers.google.com/ar/develop/java/quickstart
* Wikipedia article: https://en.wikipedia.org/wiki/ARCore

**Comment by tuxayo** [comment]
Best would be to have it as a separate library that microG would use and that other apps can integrate. So for example StreetComplete wouldn't need to use a non-libre dependency on any device. Regardless of Play Services installed or not.

Maybe there is already a libre library for some of the AR feature needed for ARCore?

**Comment by rugk** [comment]
BTW, I looked into this again and had a big surprise. Because the ARCore (Play Services) app is a separate app, which I could just install](https://play.google.com/store/apps/details?id=com.google.ar.core). According [to Exodus there are also not a lot of trackers, basically just the one you get with Google Play Services lib](https://reports.exodus-privacy.eu.org/de/reports/com.google.ar.core/latest/).
Also, you can just disallow all internet connections for the respective apps.

I tested it with the Google-developed apps "AR Elements" and "Depth Lab"… so next I am trying this with StreetComplete (it already detected that I have everything installed.)

Pixel 4 with CalyxOS 3.3.0 and microG

**Comment by spodermenpls** [comment]
I use a Sony Xperia XA2 (running Sailfish OS, which uses an emulator for AOSP-Android 10), which is unfortunately not among the supported devices cited on https://developers.google.com/ar/devices, the installation aborts immediately when I try my luck with the Aurora Store.

Therefore, I would be very keen to see a possibility to make it work. Are those "supported devices" selected by absolute necessity (meaning, the hardware of devices not mentioned forbids ARCore to work properly), or could ARCore theoretically run on more devices (with a sufficient Android API level), e.g. a Xperia XA2?

**Comment by ale5000-git** [comment]
There is probably an hard-coded list of devices inside "Google Play Services per AR" apk, so probably it need to be patched to enable support for unsupported devices.

**Comment by rugk** [comment]
…which again, likely would not help. I can totally understand that their algorithm may need some access to special sensors or so and cannot do this on any device. So it _may_ be a hardware limitation indeed.

**Comment by ale5000-git** [comment]
Not really, it doesn't work also on high-end phones that aren't certified.

**Comment by rugk** [comment]
Well I have no i idea what it needs. [Google publicly only says](https://developers.google.com/ar/develop/device-certification):
> OEM device manufacturers wishing to certify new devices may refer to the Google AR OEM Portal (only available to authorized OEMs) for a complete list of requirements.

So if anyone has any information, feel free to post it here…

**Comment by ale5000-git** [comment]
I think it simply check is device on my list thne ok, otherwise refuse to run.

**Comment by spodermenpls** [comment]
https://github.com/tomthecarrot/arcore-for-all/issues/139#issue-394048703 Can someone check if this method still works with recent releases? Reading the comments, it at least did into 2021.

**Comment by paolo-caroni** [comment]
> [tomthecarrot/arcore-for-all#139 (comment)](https://github.com/tomthecarrot/arcore-for-all/issues/139#issue-394048703) Can someone check if this method still works with recent releases? Reading the comments, it at least did into 2021.

I will confirm, but you can use adb backup only until android 11, on 12+ you need root to push the file on the correct folder. I'm actually making a config file for my phone.

**Comment by spodermenpls** [comment]
@paolo-caroni  
> I will confirm, but you can use adb backup only until android 11, on 12+ you need root to push the file on the correct folder. I'm actually making a config file for my phone.

Thank you, but as far as I can see (11 months later), implementing "ARCore" into GmsCore might be unfeasible as a whole, unfortunately. Their license file (https://github.com/google-ar/arcore-android-sdk/blob/master/LICENSE) specifies under "Section 1" seemingly the majority of binary files being restrained by the terms laid out on this page (https://developers.google.com/ar/develop/terms), which does specify for example the requirement "Your product or service must contain substantial, independent value and features beyond the Google products or services", which GmsCore may not be able to meet. Nevertheless, me not being able to use Android app features that rely on ARCore (popular among merchant chains, in particular) is still a bummer, which is why I still would love to see "ARCore" getting introduced to the otherwise great MicroG suite of stock Android service surrogates, if possible by legal means.

@mar-v-in Since you're the lead developer, and haven't said a word on the topic of "ARCore"-implementation so far (at least in this dedicated thread), I would like to hear your assessment on this - also since you're assumably dealing with Google's licensing fine print and requirements all the time.

**Comment by ale5000-git** [comment]
How for most things the best way is "Clean-room reverse engineer".

**Comment by paolo-caroni** [comment]
> @paolo-caroni
> 
> > I will confirm, but you can use adb backup only until android 11, on 12+ you need root to push the file on the correct folder. I'm actually making a config file for my phone.
> 
> Nevertheless, me not being able to use Android app features that rely on ARCore (popular among merchant chains, in particular) is still a bummer, which is why I still would love to see "ARCore" getting introduced to the otherwise great MicroG suite of stock Android service surrogates, if possible by legal means.
> 
This is not a solution, but the proprietary ARCore apk can be downloaded by AuroraStore(device spoofing), then you can add your config file (on lineageos for example uoy can root only adb) .
For pokémonGo run good on my FairPhone 3, with minor bugs (I have NOT well calibrate my sensors).
It's still proprietary, so is still wrong and immoral, but run.

About clean room seems very difficult, it's difficult to calibrate without documentations so reimplementation would be hard.

**Comment by BurgerThe** [comment]
Hi folks, I came across this issue when I was testing some things in microG. I am not sure how much of the ARCore APIs have been implemented by microG but I can confirm that the **"Live View" feature in Google Maps works**.

Tested on:
- Pixel 6a
- microG v0.3.14
- CalyxOS v7.2.1.0

Steps on how I got it working:
1. Uninstall Google Maps if already installed

2.  Install the following apps in this order:
a) Google app
b) Google Play Services for AR
c) Pixel Camera (or equivalent proprietary camera app)
d) Google Lens (unsure if this is needed, but try installing this and using before uninstalling it and testing)
e) Google Maps

3. Launch Pixel Camera and give it the permissions it needs (or equivalent camera app)

4. Force stop Google Maps

5. Start it fresh and then give it the permissions and go to Walking and you should be able to select "Live View"

<img width="472" height="1025" alt="Image" src="https://github.com/user-attachments/assets/b1caa988-a523-4077-bc81-a4a1ded1fd35" />

NOTES:
- CalyxOS marks apps installed from Aurora Store as installed from the Play Store. You may need to have a feature like that on your ROM to meet [the ARCore requirements](https://developers.google.com/ar/devices#google_play).
- I have not tested this on other devices
- I have not tested this in other apps.

Whatever the case, microG seems to support enough APIs to have ARCore working fine at least in this one specific feature that needs ARCore.
